### DISCUSSION TITLE
[PorousFlow] why my drainage is less when I increase 'flux_function' to a certain value?

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#20545] [Q&A Modules: General] [answered]
Hi guys,

so I'm using `PorousFlowPiecewiseLinearSink` for my drain BC

```
   [./drain_DRAINSRF_EOS_PITFILL]
      type = PorousFlowPiecewiseLinearSinkBEH
      variable = porepressure
      boundary = 'DRAINSRF_UG'
      pt_vals = '0 1e9'
      multipliers = '0 1e9'
      flux_function = 1
      PT_shift = 0
      v = min
      use_mobility = true
      fluid_phase = 0
     [../]
```

below is my drain surface `'DRAINSRF_UG'` generated externally (Abaqus)
![image](https://user-images.githubusercontent.com/79562439/158108309-f434f097-e2b9-4c83-a5eb-fcab4d5b5cbc.png)

and this is what I got from my BC, as you can see some of the lower parts (drives and stopes) are not being drained. This is the phreatic surface (isosurface where porepressure = 0)

![image](https://user-images.githubusercontent.com/79562439/158108423-9b5f9f87-4edc-4991-8334-d68997f36a04.png)

So I thought, increase the `flux_function` would drain more > bigger phreatic surface, but it's not the case, here is when flux_function = 1e2, I drain less compared to when flux_function = 1

![image](https://user-images.githubusercontent.com/79562439/158108638-c8aeb05c-56c3-41e2-a6b1-29509b9a35dd.png)

Not sure why this is the case?

Noted: I find this solver setting gives me the smoothes results, other standard settings dont give me anything as nice.

```
  [Executioner]
    type = Steady
    solve_type = NEWTON
    petsc_options = '-snes_converged_reason'
    petsc_options_iname = '-pc_type -pc_hypre_type -pc_hypre_boomeramg_strong_threshold -pc_hypre_boomeramg_agg_nl  -pc_hypre_boomeramg_agg_num_paths -pc_hypre_boomeramg_max_levels  -pc_hypre_boomeramg_coarsen_type   -pc_hypre_boomeramg_interp_type -pc_hypre_boomeramg_P_max -pc_hypre_boomeramg_truncfactor'
    petsc_options_value = 'hypre    boomeramg 0.5 0 4 25 Falgout ext+i 2 0.2'
    nl_rel_tol = 1e-6
    l_max_its = 300
    nl_max_its = 100
  []
```

Please let me know if you have any idea in mind, thank you!

@GiudGiud could you please help me tag any expert in this field, I know @WilkAndy is an expert on this topic

**Comment by WilkAndy** [comment, marked as answer]
I don't know the answer!

- Could it be the `v = min` in your version of the BC?  i don't know what that is doing.
- Perhaps when you run further in time, or tighten the tolerances, you'll get the expected behaviour.  Eg, when `flux_function = 100`, when using the same `nl_tol`s then you might be actually 2 orders of magnitude "less converged" than when `flux_function = 1`.
- Perhaps nonlinear effects are coming into play: if you massively drawdown a section when `flux_function = 100` and cause desaturation, then perhaps relative permeability prevents water from moving.
- If you `save_in` the fluxes, do you get more or less when `flux_function = 100`?

Well, those are some thoughts - tell us if they spark some understanding of this situation.  It's always difficult to debug huge models!  Thanks for the `petsc_options` info.

**Reply by Traiwit** [reply]
Thank you for getting back to me promptly @WilkAndy 

The reason I mention the `petsc_options` because, even with `flux_function = 1`, but different `pectsc_options`, i got totally different results.

Here is an example:  (this is an un-useable result IMO)
![image](https://user-images.githubusercontent.com/79562439/158166981-8411d36e-bda7-42e0-b938-5940d42a38e1.png)
from
```
    petsc_options_iname = '-pc_type -sub_pc_type  -pc_asm_overlap'
    petsc_options_value = ' asm      ilu           2              '
```

for `v = min`, `min `is basically the characteristic length of the element near the surface, this gives a minimal change to the solution (I just want to exactly follow Abaqus BCs)

I'm using a FullySaturated action, I also tried UnSaturated action, minimal change, also note that this is a steady-state run

I will try to address the save_in fluxes and other points tomorrow.

Thank you,
Traiwit

**Reply by WilkAndy** [reply]
This suggests to me that your model hasn't properly converged, or there are multiple solutions.  I'm struggling to see why there might be multiple solutions here, but it's quite common in nonlinear/dissipative models.  Assuming there is a unique solution, the result should not depend on choice of petsc linear solver.

Can you do some transient steps that lead up to the steady-state, eg take timestep 1s, 10s, 1e2, 1e3, ..., 1e10 or something similar?

**Reply by Traiwit** [reply]
Hi @WilkAndy,

Sure i'm following the setting from the ground water example

```
[Executioner]
  type = Transient
  solve_type = Newton
  dt = 1E6
  [TimeStepper]
    type = FunctionDT
    function = 'max(1E6, t)'
  []
  end_time = 1E12
  nl_abs_tol = 1E-13
[]
```

I will get back to you very soon

the model is massive, will take quite a bit of time

**Reply by Traiwit** [reply]
Hi @WilkAndy, so the problem was that my `nl_rel_tol` wasn't low enough, I set it to `1e-6` for the most cases in the past 

when I see the jump of R from e6 to e1 (R initial is around 5e11), I thought it wouldn't go any further
![image](https://user-images.githubusercontent.com/79562439/158711775-e35f243a-fec0-4f04-8f61-63d81f1c2f27.png)

but then I let it solve further till the system breakdown, it could reach R = e-2, then changed `nl_abs_tol` to that, and I got the results that I want (very good match Abaqus).

![image](https://user-images.githubusercontent.com/79562439/158711827-b5e06818-73a9-497b-9537-4e9ade3eacb0.png)

the wireframe is Abaqus, the surface is moose

![image](https://user-images.githubusercontent.com/79562439/158711924-5e768eab-6b53-4541-a194-83f042e498e7.png)

Thank you for your help!
Traiwit

**Reply by WilkAndy** [reply]
Phew !    I'm always stressed that something's not working in the code, so nice to know it was just some stupid tolerances :-)
